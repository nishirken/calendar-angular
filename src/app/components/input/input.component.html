<div class="host">
    <wired-input class="input" [placeholder]="placeholder" [type]="type" (input)="handleInput($event)"></wired-input>
    <div *ngIf="error" class="error">{{ error }}</div>
</div>