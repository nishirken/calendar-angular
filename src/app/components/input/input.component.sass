.host
    position: relative

.input
    box-sizing: border-box
    width: 100%

.error
    position: absolute
    bottom: -16px
    color: red
    font-size: 14px
    padding: 0 4px
    box-sizing: border-box