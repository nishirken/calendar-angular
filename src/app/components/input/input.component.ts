import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.sass'],
})
export class InputComponent {
  @Input() value = '';
  @Input() placeholder = '';
  @Input() error = '';
  @Input() type = '';
  @Input() testId?: string;
  @Input() errorTestId?: string;

  @Output() onInput = new EventEmitter<Event>();
}
