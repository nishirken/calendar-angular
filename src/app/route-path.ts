export enum RoutePath {
  Signin = 'auth/signin',
  Signup = 'auth/signup',
  Week = 'week',
}
