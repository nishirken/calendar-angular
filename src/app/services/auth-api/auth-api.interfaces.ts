import { ApiError } from '../api.interfaces';

export enum SigninErrorCode {
  CredsInvalid = 'CREDENTIALS_INVALID',
}

export type SigninApiError = ApiError<SigninErrorCode>;

export enum SignupErrorCode {
  EmailInvalid = 'EMAIL_INVALID',
  UserExists = 'USER_EXISTS',
  PasswordInvalid = 'PASSWORD_INVALID',
}

export enum PasswordInvalidErrorCode {
  TooShort = 'PASSWORD_TOO_SHORT',
  TooLong = 'PASSWORD_TOO_LONG',
  ReqChars = 'NOT_ENOUGH_REQ_CHARS',
  InvalidChars = 'INVALID_CHARS',
}

export type PasswordInvalidError =
  | {
      code: PasswordInvalidErrorCode.TooShort;
      minLength: number;
      providedLength: number;
    }
  | {
      code: PasswordInvalidErrorCode.TooLong;
      maxLength: number;
      providedLength: number;
    }
  | {
      code: PasswordInvalidErrorCode.ReqChars;
      minimumAmount: number;
      providedAmount: number;
      characterCategory: string;
    }
  | {
      code: PasswordInvalidErrorCode.InvalidChars;
      invalidCharacters: string;
    };

export type SignupApiError =
  | ApiError<SignupErrorCode.EmailInvalid, { reason: string }>
  | ApiError<SignupErrorCode.UserExists>
  | ApiError<SignupErrorCode.PasswordInvalid, PasswordInvalidError[]>;

export type AuthCreds = {
  email: string;
  password: string;
};
