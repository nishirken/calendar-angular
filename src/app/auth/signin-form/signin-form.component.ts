import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { signinFormTestIds } from 'calendar-test-ids';
import { RoutePath } from '../../app-routing.module';
import { ApiError } from '../../services/api.interfaces';
import { AuthCreds, SigninErrorCode } from '../../services/auth-api/auth-api.interfaces';
import { AuthApiService } from '../../services/auth-api/auth-api.service';
import { AuthFormComponent } from '../auth-form/auth-form.component';

const errorMessage = "Email or password are invalid";

@Component({
  selector: 'app-signin-form',
  templateUrl: '../auth-form/auth-form.component.html',
  styleUrls: ['../auth-form/auth-form.component.sass']
})
export class SigninFormComponent extends AuthFormComponent {
  constructor(private authApiService: AuthApiService, private router: Router) {
    super();
  }

  override buttonText = "Sign in";
  override hintText = "Do not have an account yet?";
  override hintLinkText = "Sign up";
  override hintLinkHref = "/auth/signup";
  override testIds = signinFormTestIds;

  override submit = (creds: AuthCreds) => {
    this.authApiService.signin(creds)
      .subscribe({
        next: () => this.router.navigate([RoutePath.Week]),
        error: (err: ApiError<SigninErrorCode>) => {
          this.emailErrorMessage = errorMessage;
          this.passwordErrorMessage = errorMessage;
        },
    });
  };

  override handleEmailChange(event: Event) {
    super.handleEmailChange(event);
    this.emailErrorMessage = '';
    this.passwordErrorMessage = '';
  }

  override handlePasswordChange(event: Event) {
    super.handlePasswordChange(event);
    this.emailErrorMessage = '';
    this.passwordErrorMessage = '';
  }
}
