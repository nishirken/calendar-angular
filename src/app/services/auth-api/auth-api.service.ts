import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, switchMap } from 'rxjs';
import { url } from '../helpers';
import { AuthCreds } from './auth-api.interfaces';

@Injectable({
  providedIn: 'root',
})
export class AuthApiService {
  constructor(private http: HttpClient) {}

  signin = (creds: AuthCreds): Observable<void> => {
    return this.http.post<void>(url('/auth/signin'), creds, {
      withCredentials: true,
    });
  };

  signup = (creds: AuthCreds): Observable<void> => {
    return this.http
      .post<void>(url('/auth/signup'), creds, { withCredentials: true })
      .pipe(switchMap(() => this.signin(creds)));
  };
}
