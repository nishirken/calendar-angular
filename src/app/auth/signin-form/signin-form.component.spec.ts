import { CUSTOM_ELEMENTS_SCHEMA, DebugElement } from '@angular/core';
import {
  ComponentFixture,
  fakeAsync,
  flushMicrotasks,
  TestBed,
  tick,
  waitForAsync,
} from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { Router } from '@angular/router';
import { RoutePath } from '../../route-path';
import { InputComponent } from '../../components/input/input.component';
import { AuthApiService } from '../../services/auth-api/auth-api.service';
import { AuthApiService as AuthApiServiceStub } from '../../services/auth-api/__mocks__/auth-api.service';
import { SigninFormComponent } from './signin-form.component';
import { RouterTestingModule } from '@angular/router/testing';
import { signinFormTestIdSelectors } from 'calendar-test-ids';
import { throwError } from 'rxjs';

describe('SigninFormComponent', () => {
  let component: SigninFormComponent;
  let fixture: ComponentFixture<SigninFormComponent>;
  let apiServiceStub: AuthApiServiceStub;
  let submitButton: HTMLButtonElement | null;
  const email = 'd@m.co';
  const password = 'Q12345';
  const router = { navigate: jest.fn() };

  beforeEach(waitForAsync(() => {
    jest.clearAllMocks();
    apiServiceStub = new AuthApiServiceStub();
    TestBed.configureTestingModule({
      declarations: [SigninFormComponent, InputComponent],
      providers: [
        { provide: AuthApiService, useValue: apiServiceStub },
        { provide: Router, useValue: router },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(SigninFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    submitButton = fixture.debugElement.query(
      By.css(signinFormTestIdSelectors.submitButton)
    ).nativeElement;
    component.emailControl.setValue(email);
    component.passwordControl.setValue(password);
  }));

  it('Renders', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('Calls signin api', () => {
    submitButton!.click();
    expect(apiServiceStub.signin).toHaveBeenCalledWith({ email, password });
  });

  it('Resets error messages after email has been changed', () => {
    component.emailErrorMessage = 'Email error';
    component.passwordErrorMessage = 'Password error';
    const emailInput = fixture.debugElement.query(
      By.css('app-input[type="email"]')
    );
    emailInput.componentInstance.onInput.next(new InputEvent('input'));
    fixture.detectChanges();
    expect(component.emailErrorMessage).toBe('');
    expect(component.passwordErrorMessage).toBe('');
  });

  it('Resets error messages after password has been changed', () => {
    component.emailErrorMessage = 'Email error';
    component.passwordErrorMessage = 'Password error';
    const emailInput = fixture.debugElement.query(
      By.css('app-input[type="password"]')
    );
    emailInput.componentInstance.onInput.next(new InputEvent('input'));
    fixture.detectChanges();
    expect(component.emailErrorMessage).toBe('');
    expect(component.passwordErrorMessage).toBe('');
  });

  it('Sets error message', () => {
    apiServiceStub.signin.mockImplementationOnce(() =>
      throwError(() => new Error())
    );
    submitButton!.click();
    fixture.detectChanges();
    const testInputError = (type: 'email' | 'password'): void => {
      expect(
        fixture.debugElement.query(By.css(`app-input[type="${type}"]`))
          .componentInstance.error
      ).toBe('Email or password are invalid');
    };
    testInputError('email');
    testInputError('password');
  });
});
