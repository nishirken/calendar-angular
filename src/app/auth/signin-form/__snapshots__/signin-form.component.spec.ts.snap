// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SigninFormComponent Renders 1`] = `
<app-signin-form
  __ngContext__="0"
  authApiService={[Function AuthApiService]}
  buttonText={[Function String]}
  emailControl={[Function FormControl2]}
  emailErrorMessage=""
  handleLinkClick={[Function Function]}
  handleSubmit={[Function Function]}
  hintLinkHref={[Function String]}
  hintLinkText={[Function String]}
  hintText={[Function String]}
  href={[Function String]}
  passwordControl={[Function FormControl2]}
  passwordErrorMessage=""
  router={[Function Object]}
  submit={[Function Function]}
  testIds={[Function Object]}
>
  <div
    class="form"
    data-test-id="signin-form"
  >
    <app-input
      class="input"
      placeholder="email"
      type="email"
    >
      <div
        class="host"
      >
        <wired-input
          class="input"
          data-test-id="signin-form-email-input"
        />
      </div>
    </app-input>
    <app-input
      class="input"
      placeholder="password"
      type="password"
    >
      <div
        class="host"
      >
        <wired-input
          class="input"
          data-test-id="signin-form-password-input"
        />
      </div>
    </app-input>
    <wired-button
      class="button"
      data-test-id="signin-form-submit-button"
    >
       Sign in 
    </wired-button>
    <div
      class="hint"
    >
       Do not have an account yet? 
      <wired-link
        class="hint-link"
        data-test-id="signin-form-link"
      >
         Sign up 
      </wired-link>
    </div>
  </div>
</app-signin-form>
`;
